### ISSUE TITLE
Feature Request: Full URLs in Builder Constructor

### ISSUE THREAD

**Issue by totev** [auth0/jwks-rsa-java#20]
Hi,

would it not be useful if the constructor of the convenience class JwkProviderBuilder would also accept a URL? Currently it is unusable for situations where the URL differs from the standard pattern (ending with "/.well-known/jwks.json"). It assumes, that the given string is only the domain-name and "/.well-known/jwks.json" is appended.

PS: I thought I write a feature request before implementing it myself and starting a pull request.

Cheers and thanks for your work,
Dobo

**Comment by lbalmaceda** [comment]
@totev We prefer the domain approach as it goes best with the normal practices. But you should be able to create a `UrlJwkProvider` from a file using the constructor that accepts an `URL` instead of a `String`. i.e. 

```java
new File(path).toURI().toURL();
//or
new URL("file:///path");
```

Maybe these ☝️ examples should be added to the javadoc? What do you think?

**Comment by totev** [comment]
@lbalmaceda thanks for the reply. This works out of the box, in fact I'am using it exactly like that in my test cases.
My Question was about the convenience class JwkProviderBuilder, which allows me to construct the  decorated JwkProviders. It only accepts the domain name and implicitly sets the location of the remote endpoint/json.

**Comment by lbalmaceda** [comment]
Ahh yes, totally! Though besides the convenience of having a builder the outcome will be the same as creating the provider yourself, like you must be doing now right? Nevertheless I will accept a PR if it:
- adds a new constructor to the builder that receives a `URL` instance.
- the current `String` constructor now calls the `URL` constructor similar to how it's being done [here](https://github.com/auth0/jwks-rsa-java/blob/master/src/main/java/com/auth0/jwk/UrlJwkProvider.java#L40). Keep in mind that the behavior should remain the same: The string should be handled as a "domain" to which the default well-known path is then appended.
- instead of keeping the `String` in a field, you now keep a `URL` instance.
- haven't tested myself but probably the `normalizeDomain` method can be ignored as the `UrlJwkProvider` constructor might be doing the sanitization already.
- add tests and javadocs

**Comment by totev** [comment]
I see your point, but just had to implement the remote keystore functionality and the provider has a non-standard path - https://acme.com/something-special/jwks.json. Now I can use the library and the available decorators (many thanks for those btw.) like so:
```		
JwkProvider provider = new UrlJwkProvider(new URL("https://acme.com/something-special/jwks.json"));
JwkProvider ratedProvider = new RateLimitedJwkProvider(provider, new BucketImpl(10, 1, TimeUnit.MINUTES));
JwkProvider cachedProvider = new GuavaCachedJwkProvider(ratedProvider, 10, 24, TimeUnit.HOURS);
return cachedProvider;
```

But can not use the much nicer builder like so:
```
JwkProvider provider = new JwkProviderBuilder("https://acme.com/something-special/jwks.json")
    .cached(10, 24, TimeUnit.HOURS)
    .rateLimited(10, 1, TimeUnit.MINUTES)
    .build();
```

Because of said implicit path appending.

I thought that the builder should have the same constructor capabilities as the UrlJwkProvider.

If you think that my use case deviates from the designed usage, that is a valid a point and this ticket should be closed. I just wanted to point out this difference.

**Comment by lbalmaceda** [comment]
Yes, I understand that it can be too verbose. But can't a PR like the one I described in my previous comment fix it? You would use the URL constructor of the `JwkProviderBuilder` instead of the current String constructor. What I meant with keeping the old behavior is that we still need to treat String as "domains" and URLs as final (but valid) urls (remote, local, files, etc). I can make it sometime before the next week unless you want to submit it.

**Comment by totev** [comment]
Oooh, I've misunderstood the "domains" part of your post.

Sure, I can gladly code it.

**Comment by darthvalinor** [comment]
Hello,

Recently, I've been thinking about the same. That's great it's raised already.

Unfortunately, this topic hasn't been updated for some time, so I made a pull-request https://github.com/auth0/jwks-rsa-java/pull/22

BR

**Comment by lbalmaceda** [comment]
@totev your use case should be covered with @darthvalinor's PR. You would be able to pass a full `URL` instance instead of the `String` domain. But now the responsibility where that URL is valid and pointing to the right JWKs file is up to you. I'll leave my review but can you give it a second look please?

**Comment by lbalmaceda** [comment]
This should be already available in `0.4.0`.
